File: 0020-valid-parentheses/0020-valid-parentheses.go
func isValid(s string) bool {
    stack := []rune{}
    for _, c := range s {
        switch c {
        case '(', '[', '{':
            stack = append(stack, c)
        case ')':
            if len(stack) == 0 || stack[len(stack)-1] != '(' {
                return false
            }
            stack = stack[:len(stack)-1]
        case ']':
            if len(stack) == 0 || stack[len(stack)-1] != '[' {
                return false
            }
            stack = stack[:len(stack)-1]
        case '}':
            if len(stack) == 0 || stack[len(stack)-1] != '{' {
                return false
            }
            stack = stack[:len(stack)-1]
        }
    }
    return len(stack) == 0
}

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    // ()
    // ([{}[]]) 
    //   ^^^^^
    // stack    c
    // (      | (
    // ([     | [
    // ([{    | {
    // ([{    | }   pop {
    // ([[    | [
    // ([[    | ]   pop [
    // ([
    // ..
    // ""     return true
    public boolean isValid(String s) {
        Map<Character, Character> map = Map.of('(', ')', '{', '}', '[', ']');
        
        Stack<Character> stack = new Stack<>();
        
        for (char c : s.toCharArray()) {
            if (map.containsKey(c)) {
                stack.push(c);
            } else if (stack.isEmpty() || c != map.get(stack.peek())) {  // { | ], // "" | ]
              return false;
            } else {
                stack.pop();
            }
        }
        return stack.isEmpty();
    }
}